actors/builtin/miner/miner_actor.go

```diff
@@ -541,88 +652,172 @@ func (a Actor) TerminateSectors(rt Runtime, params *TerminateSectorsParams) *adt
+		recoveries, err := bitfield.IntersectBitField(st.Recoveries, allDeclared)
 		if err != nil {
-			rt.Abortf(exitcode.SysErrInternal, "failed to get number of sectors")
+			rt.Abortf(exitcode.ErrIllegalState, "failed to intersect sectors with recoveries: %s", err)
+		}
+		newFaults, err := bitfield.SubtractBitField(allDeclared, recoveries)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to subtract recoveries from sectors: %s", err)
 		}
 
-		faultsMap, err := st.FaultSet.AllMap(maxAllowedFaults)
+		empty, err := newFaults.IsEmpty()
 		if err != nil {
-			rt.Abortf(exitcode.ErrIllegalState, "too many faults")
+			rt.Abortf(exitcode.ErrIllegalState, "failed to check if bitfield was empty: %s", err)
 		}
 
-		for _, sectorNumber := range dfaults {
-			sector, found, err := st.GetSector(store, abi.SectorNumber(sectorNumber))
-			if err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to load sector %v: %v", sectorNumber, err)
-			}
-			_, fault := faultsMap[uint64(sectorNumber)]
-			Assert(fault == (sector.DeclaredFaultEpoch != epochUndefined))
-			Assert(fault == (sector.DeclaredFaultDuration != epochUndefined))
-			if !found || fault {
-				continue // Ignore declaration for missing or already-faulted sector.
+		if !empty {
+			// Check new fault are really new.
+			contains, err := abi.BitFieldContainsAny(st.Faults, newFaults)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to intersect existing faults")
+			if contains {
+				// This could happen if attempting to declare a fault for a deadline that's already passed,
+				// detected and added to Faults above.
+				// Wait for the fault detection at proving period end, or submit again omitting deadlines that have
+				// passed.
+				// Alternatively, we could subtract the detected faults from new faults.
+				rt.Abortf(exitcode.ErrIllegalArgument, "attempted to re-declare fault")
 			}
 
-			storageWeightDescs = append(storageWeightDescs, asStorageWeightDesc(st.Info.SectorSize, sector))
+			// Add new faults to state and charge fee.
+			err = st.AddFaults(store, newFaults, periodStart)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add faults")
 
-			sector.DeclaredFaultEpoch = effectiveEpoch
-			sector.DeclaredFaultDuration = params.Duration
-			if err = st.PutSector(store, sector); err != nil {
-				rt.Abortf(exitcode.ErrIllegalState, "failed to update sector %v: %v", sectorNumber, err)
-			}
+			// Load info for sectors.
+			declaredFaultSectors, err = st.LoadSectorInfos(store, newFaults)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load fault sectors")
+
+			// Unlock penalty for declared faults.
+			declaredPenalty, err := unlockPenalty(store, &st, currEpoch, declaredFaultSectors, pledgePenaltyForSectorDeclaredFault)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to charge fault fee")
```

Comment:
shouldn't we check here if this is the 14th period for which it was declared faulty?

Reply:
Faults are declared only once, the transition from not fault to faulty. Then they stay faulty and the daily cron checks how old ongoing faults are. This is definitely the first period that these specific sectors have been faulty.